### ISSUE TITLE
Noise injection to generator models

### ISSUE THREAD

**Issue by ljuvela** [gillesdegottex/percivaltts#35]
I couldn't find any noise inputs to the "generator" models (spec, NM, VUV, F0). Is this intentional?

Typical GAN models inject random noise at some point in the generator model (as a source of stochasticity). Without this GANs are no longer generative, but rather just "adversarial networks". In this case, it might work to add some noise sources after the context label embedding pre-net. 

I guess dropout can also be used as a noise source, but it should also be used during test time. Haven't seen a lot of this approach in the GAN papers out there, though.

**Comment by gillesdegottex** [comment]
Indeed, there is none. I was using a random noise in parallel to the context labels at first.
Then I realized it didn't make any difference and I simply dropped it.
I suppose there is way enough implicit noise in the labels anyway (i.e. data uncorrelated to the output).
Also, this makes the generation deterministic, which is quite convenient for debugging.

**Comment by gillesdegottex** [comment]
Have you ever tried to drop this noise in your pipelines?

**Comment by ljuvela** [comment]
Haven't tried that, but it would indeed be a good sanity check. 😃 

In waveform domain generation I'm often struggling with mode collapse (or repeated wave shapes), and the human ear is very sensitive to this. I'm currently trying to get all my stochasticity from the network (no parametric noise shaping etc.), and feel that adding some noise energy is likely required.

I imagine this is less of an issue with magnitude spectra or PDD noise mask, which are in some sense parametrising the signal and controlling a noise source in vocoder synthesis.

**Comment by ljuvela** [comment]
I've also tried inputting noise and raw conditioning features at the same level (acoustic feats in my case, but the same should apply to linguistics). This didn't work too well. My hand-wavy explanation is that in this case the input layer is forced to directly mix the noise and raw conditioning, which is unlikely happening in the real data distribution. Then the network has to effectively de-mix the noise in the intermediate layers and ends up just de-noising the inputs. 

In contrast, first using an 'encoder' net to embed the conditioning, and then inputting noise and the embedding to a 'decoder' net seems to work better for me. This is not unlike what SEGAN is doing https://github.com/santi-pdp/segan

**Comment by gillesdegottex** [comment]
Exactly, with a vocoder, the noise is parametrised and built at vocoder's synthesis stage. So no big need in the generated features.

mhmmm, I never tried putting the noise at the embeded context level. What type of network are you using for this embeding? I've tried wide time windows, with some success, but I still need to make the 2 steps training with it :P
